src/entities/trade.ts

```diff
@@ -61,8 +62,35 @@ export class Trade<TInput extends Currency, TOutput extends Currency, TTradeType
         outputAmount,
       })
     }
-
     this.tradeType = tradeType
+
+    // each route must have the same input and output currency
+    const inputCurrency = this.swaps[0].inputAmount.currency
+    const outputCurrency = this.swaps[0].outputAmount.currency
+    invariant(
+      this.swaps.every(({ route }) => inputCurrency.wrapped.equals(route.input.wrapped)),
+      'INPUT_CURRENCY_MATCH'
+    )
+    invariant(
+      this.swaps.every(({ route }) => outputCurrency.wrapped.equals(route.output.wrapped)),
+      'OUTPUT_CURRENCY_MATCH'
+    )
+
+    // pools must be unique inter protocols
+    const numPools = this.swaps.map(({ route }) => route.pools.length).reduce((total, cur) => total + cur, 0)
+    const poolAddressSet = new Set<string>()
+    for (const { route } of this.swaps) {
+      for (const pool of route.pools) {
+        if (route.protocol === Protocol.V3) {
+          const poolV3 = pool as Pool
+          poolAddressSet.add(Pool.getAddress(poolV3.token0, poolV3.token1, poolV3.fee))
+        } else {
+          const pair = pool as Pair
```

Comment:
FYI the casts here shouldn't be needed. Typescript should automatically narrow the type to type Pool since protocol is V3:

e.g. https://github.com/Uniswap/routing-api/blob/main/lib/handlers/quote/quote.ts#L227-L229

You would need to add another if statement for V2 below to get the same for pair